src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,241 @@
+    auto value_column = std::dynamic_pointer_cast<const ValueColumn<T>>(column);
+    if (value_column) {
+      // since ValueColumns are not sorted, just search through them linearly
+      const auto& values = value_column->values();
+      for (auto chunk_offset = ChunkOffset(0); chunk_offset < values.size(); ++chunk_offset) {
+        if (t_comparator(values[chunk_offset], t_value)) {
+          pos_list->push_back(RowID{chunk_id, chunk_offset});
+        }
+      }
+      continue;
+    }
+
+    auto dict_column = std::dynamic_pointer_cast<const DictionaryColumn<T>>(column);
+    DebugAssert(dict_column,
+                "Unknown Column Type in Table Scan");  // make sure the above is the case; only in debug mode
+
+    const auto dictionary = dict_column->dictionary();
+
+    // depending on the operator we need to select a proper ValueID
+    // which accurately represents the Value we are searching for
+    // (keep in mind it can be the case that the value we are searching for is not contained)
+
+    bool value_contained = std::binary_search(dictionary->cbegin(), dictionary->cend(), t_value);
+
+    if (!value_contained) {
+      if (scan_type == ScanType::OpEquals) {
+        // no value can be equal
+        continue;
+      }
+      if (scan_type == ScanType::OpNotEquals) {
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+    }
+
+    ValueID comp_value;
+    switch (scan_type) {
+      case ScanType::OpEquals:
+      case ScanType::OpNotEquals:
+      case ScanType::OpLessThan:
+      case ScanType::OpGreaterThanEquals:
+        comp_value = dict_column->lower_bound(t_value);
+        break;
+      case ScanType::OpLessThanEquals:
+      case ScanType::OpGreaterThan:
+        comp_value = dict_column->upper_bound(t_value);
+        break;
+    }
+
+    if (comp_value == ValueID{0u}) {
+      if (scan_type == ScanType::OpLessThanEquals || scan_type == ScanType::OpLessThan) {
+        // value is smaller than all the values in the column, there are no matches
+        // we can go to the next chunk
+        continue;
+      }
+      if (scan_type == ScanType::OpGreaterThan || scan_type == ScanType::OpGreaterThanEquals) {
+        // every value is greater than the search value - add every position
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+    }
+
+    // if upper_bound goes past the end
+    if (comp_value == INVALID_VALUE_ID) {
+      if (scan_type == ScanType::OpLessThanEquals || scan_type == ScanType::OpLessThan) {
+        // value is greater than all the values in the column
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+      if (scan_type == ScanType::OpGreaterThan || scan_type == ScanType::OpGreaterThanEquals) {
+        // every value is smaller than the search value
+        continue;
+      }
+    }
+
+    if (comp_value != ValueID{0} && comp_value != INVALID_VALUE_ID &&
```

Comment:
The first two conditions are not really necessary, because if they were true, `continue;` would already have been called above.

Reply:
Oh, I see this is where you fix the comp_value when using upper_bound. A little bit confusing.

I guess from @stephde's comment you can also see that it's not quite obvious how the control flow works here for different operations. It might introduce some duplicated code but I think it could help to introduce separate methods handling only one of the operators.